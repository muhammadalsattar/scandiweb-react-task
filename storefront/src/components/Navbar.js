import React from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import CartOverlay from "./CartOverlay";
import {setDefaultCategory} from "../actions/categories";
import {setDefaultCurrency} from "../actions/currencies";

export class Navbar extends React.Component {
    componentDidMount() {
        document.addEventListener("click", (e) => {
            if(document.querySelector(".cart-overlay").contains(e.target)){
                document.querySelector(".cart-overlay").removeAttribute("hidden");
                document.querySelector(".page-body").classList.add("faded")
            }
            else if (!e.target.parentNode?.classList.contains("cart-img")) {
                document.querySelector(".cart-overlay").setAttribute("hidden", "true");
                document.querySelector(".page-body").classList.remove("faded")
            }
            else if (!e.target.classList.contains("currency-icon")) {
                document.querySelector(".currency-dropdown").setAttribute("hidden", "true");
                document.querySelector(".currency-drop-icon").style.transform = "rotate(0deg)";
            }

        })
    }
    setDefaultCurrency = (e) => {
        e.preventDefault();
        const currency = this.props.currencies.find(currency => currency.symbol === e.target.value);
        this.props.setDefaultCurrency(currency);
    }
    openCart = () => {
        document.querySelector(".cart-overlay").removeAttribute("hidden");
        document.querySelector(".page-body").classList.add("faded");
    }
    toggleCurrencyDropdown = () => {
        document.querySelector(".currency-dropdown").removeAttribute("hidden");
        document.querySelector(".currency-drop-icon").style.transform = "rotate(180deg)";
    }
    render() {
        return (
            <nav className="navbar">
                <div className="categories">
                    {this.props.categories.map(category => {
                        return(category.name === this.props.defaultCategory.name ? <button key={category.name} className="category active">{category.name}</button> : <button className="category" key={category.name} onClick={()=>{this.props.setDefaultCategory(category)}}>{category.name}</button>)
                    })}
                </div>
                <div className="brand-logo">
                    <Link to="/"><img src={process.env.PUBLIC_URL +'/logo.png'} alt="logo-icon"></img></Link>
                </div>
                <div className="dropdowns">
                    <div className="currency-div">
                        <button onClick={this.toggleCurrencyDropdown} className="currency-icon">{this.props.defaultCurrency.symbol}<span className="currency-drop-icon">&#8743;</span></button>
                        <ul className="currency-dropdown" hidden>
                            {this.props.currencies.map(currency => <li key={currency.symbol}><button value={currency.symbol} onClick={this.setDefaultCurrency}>{currency.symbol} {currency.label}</button></li>)}
                        </ul>
                    </div>
                    <div className="cart-div">
                        <button className="cart-img" onClick={this.openCart}><img src={process.env.PUBLIC_URL+'/cart.png'} alt='cart-img'></img></button>
                        <button className="cart-count">{this.props.cart.length}</button>
                        <CartOverlay/>
                    </div>
                </div>
            </nav>
        )
    }
}

const mapDispatchToProps = dispatch => {
    return {
        setDefaultCategory: (category) => {
            dispatch(setDefaultCategory(category));
        },
        setDefaultCurrency: (currency) => {
            dispatch(setDefaultCurrency(currency));
        }
    }
}

const mapStateToProps = state => {
    return {
        categories: state.categories.categories,
        currencies: state.currencies.currencies,
        defaultCategory: state.categories.defaultCategory,
        defaultCurrency: state.currencies.defaultCurrency,
        cart: state.cart.cart,
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(Navbar);