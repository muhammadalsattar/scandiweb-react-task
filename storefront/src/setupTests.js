const currencies = [
    {
        label: 'USD',
        symbol: '$',
    },
    {
        label: 'GBP',
        symbol: 'Â£',
    },
];
const categories = [
    {
        name: 'category1',
    },
    {
        name: 'category2',
    },
];
const attributes = [
    {
        id: 'Size',
        name: 'Size',
        type: 'text',
        items: [
            {
                id: 40,
                value: '40',
                displayValue: '40',
            },
        ],
    },
    {
        id: 'Color',
        name: 'Color',
        type: 'swatch',
        items: [
            {
                id: 'red',
                value: 'red',
                displayValue: 'Red',
            },
        ],
    },
];
const gallery = [
    'https://dummyimage.com/600x400/000/fff',
    'https://dummyimage.com/600x400/000/ggg',
    'https://dummyimage.com/600x400/000/hhh',
];
const prices = [
    {
        amount: 518.47,
        currency: currencies[0],
    },
    {
        amount: 372.67,
        currency: currencies[1],
    },
];
const products = [
    {
        id: '1',
        name: 'product1',
        brand: 'brand1',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '40' }],
    },
    {
        id: '2',
        name: 'product2',
        brand: 'brand2',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'red' }],
    },
    {
        id: '3',
        name: 'product3',
        brand: 'brand3',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '41' }],
    },
    {
        id: '4',
        name: 'product4',
        brand: 'brand4',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'orange' }],
    },
    {
        id: '5',
        name: 'product5',
        brand: 'brand5',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '42' }],
    },
    {
        id: '6',
        name: 'product6',
        brand: 'brand6',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'green' }],
    },
];

export const initialState = {
    currencies,
    categories,
    products,
};
