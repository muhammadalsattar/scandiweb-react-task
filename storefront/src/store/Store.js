import { configureStore, combineReducers } from '@reduxjs/toolkit';
import cartReducer from '../reducers/cartReducer';
import productsReducer from '../reducers/productsReducer';
import categoriesReducer from '../reducers/categoriesReducer';
import currencyReducer from '../reducers/currencyReducer';

const rootReducer = combineReducers({
    cart: cartReducer,
    products: productsReducer,
    categories: categoriesReducer,
    currencies: currencyReducer,
});
// creating store and combining reducers
const Store = configureStore({
    reducer: rootReducer,
});

export default Store;
