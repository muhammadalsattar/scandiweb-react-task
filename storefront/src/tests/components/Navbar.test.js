import React from "react";
import {render, screen} from "@testing-library/react";
import { Provider } from "react-redux";
import {BrowserRouter as Router} from "react-router-dom";
import Store from "../../store/Store.js";
import {Navbar} from "../../components/Navbar.js";
import { initialState } from "../../setupTests.js"

test("Navbar renders correctly", () => {
    const { undefined, container } = render(
        <Provider store={Store}>
        <Router><Navbar categories={initialState.categories} currencies={initialState.currencies} cart={initialState.cart} defaultCurrency={initialState.currencies[0]} defaultCategory={initialState.categories[0]}/></Router>
        </Provider>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText(/all/i)).toBeInTheDocument();
});