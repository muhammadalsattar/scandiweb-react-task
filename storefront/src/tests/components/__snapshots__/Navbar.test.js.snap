// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<div>
  <nav
    class="navbar"
  >
    <div
      class="categories"
    >
      <button
        class="category active"
      >
        all
      </button>
      <button
        class="category"
      >
        clothes
      </button>
      <button
        class="category"
      >
        tech
      </button>
    </div>
    <div
      class="brand-logo"
    >
      <a
        href="/"
      >
        <img
          alt="logo-icon"
          src="/logo.png"
        />
      </a>
    </div>
    <div
      class="dropdowns"
    >
      <div
        class="currency-div"
      >
        <button
          class="currency-icon"
        >
          $
          <span
            class="currency-drop-icon"
          >
            ∧
          </span>
        </button>
        <ul
          class="currency-dropdown"
          hidden=""
        >
          <li>
            <button
              value="$"
            >
              $
               
              USD
            </button>
          </li>
          <li>
            <button
              value="£"
            >
              £
               
              GBP
            </button>
          </li>
          <li>
            <button
              value="A$"
            >
              A$
               
              AUD
            </button>
          </li>
          <li>
            <button
              value="¥"
            >
              ¥
               
              JPY
            </button>
          </li>
          <li>
            <button
              value="₽"
            >
              ₽
               
              RUB
            </button>
          </li>
        </ul>
      </div>
      <div
        class="cart-div"
      >
        <button
          class="cart-img"
        >
          <img
            alt="cart-img"
            src="/cart.png"
          />
        </button>
        <button
          class="cart-count"
        >
          2
        </button>
        <div
          class="cart-overlay"
          hidden=""
        >
          <p>
            <b>
              My Bag, 
            </b>
            0
             items
          </p>
          <div
            class="cart-total"
          >
            <p>
              Total
            </p>
            <p>
              0.00
            </p>
          </div>
          <div
            class="bag-checkout"
          >
            <div
              class="view-bag"
            >
              <a
                href="/cart"
              >
                <button>
                  View Bag
                </button>
              </a>
            </div>
            <div
              class="checkout"
            >
              <button>
                Check out
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
